dev/src/DocFx/Node/ClassNode.php

```diff
@@ -0,0 +1,170 @@
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Dev\DocFx\Node;
+
+use SimpleXMLElement;
+
+class ClassNode
+{
+    use DocblockTrait;
+    use NameTrait;
+
+    private $childNode;
+
+    public function __construct(
+        private SimpleXMLElement $xmlNode
+    ) {}
+
+    public function isProtobufEnumClass(): bool
+    {
+        if ($this->getExtends()) {
+            return false;
+        }
+
+        if (empty($this->xmlNode->docblock)) {
+            return false;
+        }
+
+        if (empty($this->xmlNode->docblock->{'long-description'})) {
+            if (empty($this->xmlNode->docblock->{'description'})) {
+                return false;
+            }
+            $description = $this->xmlNode->docblock->{'description'};
+        } else {
+            $description = (string) $this->xmlNode->docblock->{'long-description'};
+        }
+
+        // check that last line of long-description starts with "Protobuf type..."
+        $descriptionParts = explode("\n", $description);
+        $lastDescriptionLine = array_pop($descriptionParts);
+        return 0 === strpos($lastDescriptionLine, 'Protobuf type');
+    }
+
+    public function isGapicEnumClass(): bool
+    {
+        // returns true if the class extends \Google\Protobuf\Internal\Message
+        return false !== strpos($this->getNamespace(), '\Enums\\');
+    }
+
+    public function isServiceClass(): bool
+    {
+        // returns true if the class extends a generated GAPIC client
+        if ($extends = $this->getExtends()) {
+            return 'GapicClient' === substr($extends, -11);
+        }
+        return false;
+    }
+
+    public function getStatus(): string
+    {
+        if (!$this->xmlNode->docblock) {
+            return '';
+        }
+
+        foreach ($this->xmlNode->docblock->tag as $tag) {
+            if ((string) $tag['name'] === 'deprecated') {
+                return 'deprecated';
+            }
+        }
+
+        return '';
+    }
+
+    public function getExtends(): string
+    {
+        return (string) $this->xmlNode->extends;
+    }
+
+    public function isInternal(): bool
+    {
+        if (!$this->xmlNode->docblock) {
+            return '';
+        }
+
+        foreach ($this->xmlNode->docblock->tag as $tag) {
+            if ((string) $tag['name'] === 'internal') {
+                return 'true';
+            }
+        }
+
+        return false;
+    }
+
+    public function getMethods(): array
+    {
+        $methods = [];
+        foreach ($this->xmlNode->method as $methodNode) {
+            $method = new MethodNode($methodNode);
+            if ($method->isPublic() && !$method->isInherited()) {
+                $methods[] = $method;
+            }
+        }
+
+        if ($this->childNode) {
+            foreach ($this->childNode->getMethods() as $childMethod) {
+                $childMethod->setParentNode($this);
+                $methods[] = $childMethod;
```

Comment:
Do these child nodes need to follow the same restrictions as other nodes i.e. `isPublic && !isInherited`?

Reply:
That's already handled explicitly... since `getMethods` is called _on the child node_ (which is a `ClassNode` also), the `isPublic` and `isInherited` is called inside `$this->childNode->getMethods()`